src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp

```diff
@@ -0,0 +1,175 @@
+
+  // clang-tidy: google-runtime-references
+  // clang-tidy: cppcoreguidelines-owning-memory,-warnings-as-errors
+  WRAPPED_PUPable_decl_template(QuaternionFunctionOfTime<MaxDeriv>);  // NOLINT
+
+  void reset_expiration_time(double next_expiration_time) noexcept;
+
+  /// Returns domain of validity for the function of time
+  std::array<double, 2> time_bounds() const noexcept override {
+    return {{stored_quaternions_and_times_.front().time, expiration_time_}};
+  }
+
+  /// Updates the `MaxDeriv`th derivative of the omega piecewisepolynomial at
+  /// the given time, then updates the stored quaternions.
+  ///
+  /// `updated_max_deriv` is a datavector of the `MaxDeriv`s for each component.
+  /// `next_expiration_time` is the next expiration time.
+  void update(double time_of_update, DataVector updated_max_deriv,
+              double next_expiration_time) noexcept;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) override;
+
+  /// Returns the quaternion at an arbitrary time `t`.
+  std::array<DataVector, 1> func(const double t) const noexcept override {
+    return quat_func(t);
+  }
+
+  /// Returns the quaternion and its first derivative at an arbitrary time `t`.
+  std::array<DataVector, 2> func_and_deriv(
+      const double t) const noexcept override {
+    return quat_func_and_deriv(t);
+  }
+
+  /// Returns the quaternion and the first two derivatives at an arbitrary
+  /// time `t`.
+  std::array<DataVector, 3> func_and_2_derivs(
+      const double t) const noexcept override {
+    return quat_func_and_2_derivs(t);
+  }
+
+  /// Returns the quaternion at an arbitrary time `t`.
+  std::array<DataVector, 1> quat_func(double t) const noexcept;
+
+  /// Returns the quaternion and its first derivative at an arbitrary time `t`.
+  std::array<DataVector, 2> quat_func_and_deriv(double t) const noexcept;
+
+  /// Returns the quaternion and the first two derivatives at an arbitrary
+  /// time `t`.
+  std::array<DataVector, 3> quat_func_and_2_derivs(double t) const noexcept;
+
+  /// Returns stored omega at an arbitrary time `t`.
+  std::array<DataVector, 1> omega_func(const double t) const noexcept {
+    return omega_f_of_t_.func(t);
+  }
+
+  /// Returns stored omega and its first derivative at an arbitrary time `t`.
+  std::array<DataVector, 2> omega_func_and_deriv(
+      const double t) const noexcept {
+    return omega_f_of_t_.func_and_deriv(t);
+  }
+
+  /// Returns stored omega and the first two derivatives at an arbitrary
+  /// time `t`.
+  std::array<DataVector, 3> omega_func_and_2_derivs(
+      const double t) const noexcept {
+    return omega_f_of_t_.func_and_2_derivs(t);
+  }
+
+ private:
+  std::vector<FunctionOfTimeHelpers::StoredInfo<1, false>>
+      stored_quaternions_and_times_;
+
+  domain::FunctionsOfTime::PiecewisePolynomial<MaxDeriv> omega_f_of_t_;
+  double expiration_time_{std::numeric_limits<double>::lowest()};
```

Comment:
Technically they could be different, with the requirement that the domain of validity of the quaternions is contained within the domain of validity of the internal PiecewisePolynomial. However, I don't think we want that because then we'd have to constantly check that the expiration times are correct relative to each other. They should just be the same all the time.

I'll get rid of the member variable and just use the internal PiecewisePolynomial time bounds and expiration time.